:zap: Tuning gRPC server

mhkarimi1383 opened the pull request:
Improve gRPC server to only server Services when needed and set number of workers to 1 to match other provisioners

mhkarimi1383 commented:
@tiagolobocastro snappshotter should be in controller and we should use remote fn to run them on nodes

it may pick a wrong Resource from a wrong node

tiagolobocastro commented:
> @tiagolobocastro snappshotter should be in controller and we should use remote fn to run them on nodes
> 
> it may pick a wrong Resource from a wrong node

we're running in a different mode than usual here, where controller actually runs on the node
I think I have a diagram somewhere here in a github comment

also for snapshots we're running the distributed snapshotting for local volumes: <https://github.com/kubernetes-csi/external-snapshotter?tab=readme-ov-file#distributed-snapshotting>

mhkarimi1383 commented:
But snapshot methods are in the Controller part, I have lvm local pv, and it was in the controller

tiagolobocastro commented:
> But snapshot methods are in the Controller part, I have lvm local pv, and it was in the controller

Yeah we run the unified deployment mode, it's not the same as lvm local pv.
I've mentioned this before: <https://github.com/openebs/rawfile-localpv/pull/28#discussion_r1974043841>

mhkarimi1383 commented:
Will check it out

mhkarimi1383 commented:
Since this is a thing in csi proto and I can't change it I will leave node plugin to serve controller things too for now

tiagolobocastro commented on rawfile/rawfile.py:
```diff
@@ -61,18 +61,31 @@ def cli(
+        enable_metrics=enable_metrics,
+        plugin_type=plugin_type,
+    )
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=1))
```
why is this being modified to 1 worker?

mhkarimi1383 replied:
It was to match others, but We should have a better value, I have the same question that why it is 10 right now

tiagolobocastro replied:
others as in?
I'm not sure exactly why 10 here, but on mayastor for example we have a default value of 10 which allows up to 10 concurrent csi operations

mhkarimi1383 replied:
I saw LVM I think

mhkarimi1383 replied:
I will default it to 10 then I will make it configurable

mhkarimi1383 commented:
@Mergifyio squash

mhkarimi1383 commented:
@Mergifyio squash

mhkarimi1383 committed:
:zap: Tuning gRPC server (#164)
:zap: Tuning gRPC server

Improve gRPC server to only server Services when needed
and set number of workers to 1 to match other provisioners

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* Merge branch 'develop' into grpc-server-optimize

Signed-off-by: Muhammed Hussein karimi <info@karimi.dev>

* :wrench: Make gRPC workers configurable and default it to `10`

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

* Merge branch 'develop' into grpc-server-optimize

* Merge branch 'develop' into grpc-server-optimize

mhkarimi1383 commented:
@Mergifyio queue
